src/3d/qgs3dutils.cpp

```diff
@@ -0,0 +1,223 @@
+  if ( maxError <= 0 || tileResolution <= 0 || tile0width <= 0 )
+    return 0;  // invalid input
+
+  // derived from:
+  // tile width [map units] = tile0width / 2^zoomlevel
+  // tile error [map units] = tile width / tile resolution
+  // + re-arranging to get zoom level if we know tile error we want to get
+  double zoomLevel = -log( tileResolution * maxError / tile0width ) / log( 2 );
+  return round( zoomLevel );  // we could use ceil() here if we wanted to always get to the desired error
+}
+
+QString Qgs3DUtils::altClampingToString( AltitudeClamping altClamp )
+{
+  switch ( altClamp )
+  {
+    case AltClampAbsolute: return QStringLiteral( "absolute" );
+    case AltClampRelative: return QStringLiteral( "relative" );
+    case AltClampTerrain: return QStringLiteral( "terrain" );
+    default: Q_ASSERT( false ); return QString();
+  }
+}
+
+
+AltitudeClamping Qgs3DUtils::altClampingFromString( const QString &str )
+{
+  if ( str == "absolute" )
+    return AltClampAbsolute;
+  else if ( str == "terrain" )
+    return AltClampTerrain;
+  else   // "relative"  (default)
+    return AltClampRelative;
+}
+
+
+QString Qgs3DUtils::altBindingToString( AltitudeBinding altBind )
+{
+  switch ( altBind )
+  {
+    case AltBindVertex: return QStringLiteral( "vertex" );
+    case AltBindCentroid: return QStringLiteral( "centroid" );
+    default: Q_ASSERT( false ); return QString();
+  }
+}
+
+
+AltitudeBinding Qgs3DUtils::altBindingFromString( const QString &str )
+{
+  if ( str == "vertex" )
+    return AltBindVertex;
+  else  // "centroid"  (default)
+    return AltBindCentroid;
+}
+
+
+void Qgs3DUtils::clampAltitudes( QgsLineString *lineString, AltitudeClamping altClamp, AltitudeBinding altBind, const QgsPoint &centroid, float height, const Qgs3DMapSettings &map )
+{
+  for ( int i = 0; i < lineString->nCoordinates(); ++i )
+  {
+    float terrainZ = 0;
+    if ( altClamp == AltClampRelative || altClamp == AltClampTerrain )
+    {
+      QgsPointXY pt;
+      if ( altBind == AltBindVertex )
+      {
+        pt.setX( lineString->xAt( i ) );
+        pt.setY( lineString->yAt( i ) );
+      }
+      else
+      {
+        pt.set( centroid.x(), centroid.y() );
+      }
+      terrainZ = map.terrainGenerator()->heightAt( pt.x(), pt.y(), map );
+    }
+
+    float geomZ = 0;
+    if ( altClamp == AltClampAbsolute || altClamp == AltClampRelative )
+      geomZ = lineString->zAt( i );
+
+    float z = ( terrainZ + geomZ ) * map.terrainVerticalScale() + height;
+    lineString->setZAt( i, z );
+  }
+}
+
+
+bool Qgs3DUtils::clampAltitudes( QgsPolygonV2 *polygon, AltitudeClamping altClamp, AltitudeBinding altBind, float height, const Qgs3DMapSettings &map )
+{
+  if ( !polygon->is3D() )
+    polygon->addZValue( 0 );
+
+  QgsPoint centroid;
+  if ( altBind == AltBindCentroid )
+    centroid = polygon->centroid();
+
+  QgsCurve *curve = const_cast<QgsCurve *>( polygon->exteriorRing() );
+  QgsLineString *lineString = dynamic_cast<QgsLineString *>( curve );
```

Comment:
Geometries get segmented earlier - QgsPolygonV2 instances are being passed to this method.

We should probably improve the API so that exteriorRing() of the polygon class returns QgsLineString and not a more generic QgsCurve.